### DISCUSSION TITLE
Suggestions for a test framework to use with the project

### DISCUSSION THREAD

**Discussion by jarmos-san** [developersIndia/developersIndia#100] [Ideas]
If you might've noticed, the project doesn't have **ANY** tests right now :grin: . I know, for those of you who're experienced enough out there, it sounds like a ticking bomb ready to explode any moment! As such we would like to discuss which frameworks should we choose to go with.

The official Next.js docs recommends the following frameworks for test Next.js-powered apps:

- [Cypress](https://www.cypress.io/)
- [Playwright](https://playwright.dev/)
- [Jest](https://jestjs.io/) + [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/)

I reviewed each of those frameworks & came to the following conclusions:

1. Cypress is an end-to-end testing framework & it's syntax relies heavily on [Callback functions](https://developer.mozilla.org/en-US/docs/Glossary/Callback_function) as such if not taken care of well enough, it could lead to [callback hell](http://callbackhell.com/) pretty easily. As a community-maintained project which receives contributions from a wide array of individuals with varying levels of expertise & knowledge, having to review code with too many callbacks can be painful.
2. Playwright is also an end-to-end framework & its syntax looks pretty similar to that of Cypress with a caveat. It's syntax relies on the significantly more readable & commonly used [Async-Await](https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Asynchronous/Async_await) keywords. It also appears the framework treats TypeScript & running the tests in CI environments as first-class citizens so that's a major plus point.
3. Jest & React Testing Library are the most widely used testing tools for React-based projects. But the drawback of using them is the need to setup additional tools like [Babel](https://babeljs.io/) since Jest doesn't support TypeScript out-of-the-box.

That said, if you feel either of these test frameworks will suit the project's requirements then please feel free to let us know in the comments below! Additionally, let us know if we could use other test framework that we might've missed out on as well.

**Comment by jarmos-san** [comment]
I did some more customary looking around on this topic & realised Playwright was created by the same guys behind the [Puppeteer](https://github.com/puppeteer/puppeteer) project. The only difference being, while the former is backed by Microsoft, the later was created at Google. Also Playwright supports better out-of-the-box cross-browser compatibility while Puppeteer only supported the FOSS Chromium browser!

I also noticed, Playwright is used to test client-side functionalities, integration tests & component testing [isn't supported yet](https://github.com/microsoft/playwright/issues/7148) (**_need further clarification on this statement_**) .

**Comment by nirmalhk7** [comment]
From what I've noticed, it takes _extremely_ long for CI to run integration tests (wiith Playwright or Cypress, and with caching too). Despite the Babel setup, I think we should prioritize unit tests more than integration tests. Its simple, its fast and it works. Given that our website will only be static for a foreseeable future, I dont think integration tests make too much sense.

**Reply by jarmos-san** [reply]
Yes, you're right about Playright & Cypress tests running for a long time. That happens because those tools are meant to be used for end-to-end (E2E) testing which correlates to the user experience as much as possible. I doubt we need that any time soon since we're working with just a minimal static page.

But for the long run there should be a robust test suite in place so that we can confidently release new feature enhancements to the website as & when possible. Not doing so is an irresponsibility but we should start working on writing _good_ unit tests as part of our [Next milestones](https://github.com/developersIndia/website/milestone/2).
